aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  has_free_variables_ = false;
+
+  for (const TypePtr type : types_) {
+    if (type->kind() == NoneType::Kind) {
+      can_hold_none_ = true;
+    }
+    if (type->hasFreeVariables()) {
+      has_free_variables_ = true;
+    }
+  }
+
+}
+
+UnionTypePtr UnionType::create(std::vector<TypePtr> types) {
+  auto union_type = new UnionType(std::move(types));
+  if (auto optional_type = union_type->to_optional()) {
+    return OptionalType::create(std::move(*optional_type));
+  }
+  return UnionTypePtr(std::move(union_type));
+}
+
+bool UnionType::operator==(const Type& rhs) const {
+  if (auto union_rhs = rhs.cast<UnionType>()) {
+    return types_ == union_rhs->types_;
+  }
+  if (auto optional_rhs = rhs.cast<OptionalType>()) {
+    auto optional_lhs = this->to_optional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
+  }
+  return false;
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors. We need to use
+    // `std::equal` with `TypeEqual` for optimal performance; do not
+    // change this to `operator==`!
+    if (std::equal(this->types_.begin(), this->types_.end(),
+                      union_rhs->types().begin(), union_rhs->types().end(),
+                      TypeEqual())) {
+      return true;
+    }
+    // Final O(n^2) check since the vector equality comparison above
+    // won't handle subtyping relationships
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return this_type->isSubtypeOf(union_rhs);
+    });
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
+      (types_[0]->isSubtypeOf(inner) ^ types_[1]->isSubtypeOf(inner));
+  } else if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union[";
+  for (size_t i = 0; i < types().size(); ++i) {
+    if (i > 0) {
+      ss << ", ";
+    }
+    ss << types()[i]->str();
+  }
+  ss << "]";
+  return ss.str();
+}
+
+UnionTypePtr UnionType::union_of(std::vector<TypePtr>& rhs_types) const {
+  return union_of({rhs_types});
+}
+
+UnionTypePtr UnionType::union_of(const UnionTypePtr rhs) const {
+  std::vector<TypePtr> new_types = this->types().vec();
+  new_types.insert(new_types.end(), rhs->types().begin(), rhs->types().end());
+  return UnionType::create(std::move(new_types));
+}
+
+UnionTypePtr UnionType::intersection_of(std::vector<TypePtr>& rhs_types) const {
+  return intersection_of({rhs_types});
+}
+
+UnionTypePtr UnionType::intersection_of(const UnionTypePtr rhs) const {
```

Comment:
I talked with @gmagogsfm offline about this, and we agreed that it would be best to delete `union_of` and `intersection_of` (sadly). If we're all arguing about what "intersection" means when you take subtyping relationships into account, then I don't trust that future users of this API will be able to tell what it means, either.